packages/tasit-action/src/contract/Action.js

```diff
@@ -7,54 +7,59 @@ import ConfigLoader from "../ConfigLoader";
 export class Action extends Subscription {
   #provider;
   #signer;
-  #rawTx;
+  #rawAction;
   #tx;
   #txConfirmations;
   #timeout;
   #lastConfirmationTime;
 
-  constructor(rawTx, provider, signer) {
+  constructor(rawAction, provider, signer) {
     // Provider implements EventEmitter API and it's enough
     //  to handle with transactions events
     super(provider);
 
     const { events } = ConfigLoader.getConfig();
     const { timeout } = events;
 
-    this.#rawTx = rawTx;
+    this.#rawAction = rawAction;
     this.#signer = signer;
     this.#timeout = timeout;
     this.#provider = provider;
     this.#txConfirmations = 0;
   }
 
-  _toRaw = async () => {
-    return await this.#rawTx;
+  _toRaw = () => {
+    return this.#rawAction;
   };
 
   #signAndSend = async () => {
-    // TODO: Go deep on gas handling.
-    // Without that, VM returns a revert error instead of out of gas error.
-    // See: https://github.com/tasitlabs/TasitSDK/issues/173
-    //
-    // This command isn't enough
-    // const gasLimit = await this.#provider.estimateGas(this.#rawTx);
-    const gasParams = {
-      gasLimit: 7e6,
-      gasPrice: 1e9,
-    };
+    try {
+      // Note: Resolving promise if the Action was created using a async rawTx
+      let rawTx = await this.#rawAction;
 
-    const nonce = await this.#provider.getTransactionCount(
-      this.#signer.address
-    );
+      const nonce = await this.#provider.getTransactionCount(
+        this.#signer.address
+      );
+      const gasPrice = await this.#provider.getGasPrice();
+      const network = await this.#provider.getNetwork();
+      const chainId = network.chainId;
+      let { value } = rawTx;
+      value = !value ? 0 : value;
 
-    let rawTx = await this.#rawTx;
+      rawTx = { ...rawTx, nonce, gasPrice, chainId, value };
 
-    rawTx = { ...rawTx, nonce, ...gasParams };
+      // TODO: Go deep on gas handling.
+      // Without that, VM returns a revert error instead of out of gas error.
+      // See: https://github.com/tasitlabs/TasitSDK/issues/173
+      //
+      // This command isn't enough
+      //const gasLimit = await this.#provider.estimateGas(rawTx);
+      const gasLimit = 7e6;
 
-    const signedTx = await this.#signer.sign(rawTx);
+      this.#rawAction = { ...rawTx, gasLimit };
```

Comment:
As far as I know, the gas limit estimation will work better if the `tx` is more closest to what will be broadcasted.

Reply:
An example: https://github.com/ethers-io/ethers.js/issues/300#issuecomment-429477667

Reply:
Oh - what I thought I was suggesting was just an stylistic choice that wouldn't impact functionality.

`rawTx = { ...rawTx, nonce, gasPrice, chainId, value };`

compared with

`this.#rawAction = { ...rawTx, gasLimit };`

Reply:
> Oh - what I thought I was suggesting was just an stylistic choice that wouldn't impact functionality.
> 
> `rawTx = { ...rawTx, nonce, gasPrice, chainId, value };`
> 
> compared with
> 
> `this.#rawAction = { ...rawTx, gasLimit };`

Hmmm, got it now.
Could you take a look on the code after latest commits? I think that could be deprecated since I've extracted this part of the code to a function.

Reply:
Ah ok, will do!

Reply:
Yep, deprecated now. Resolving this. Thanks!